### ISSUE TITLE
Unable to add Microsoft account

### ISSUE THREAD

**Issue by jkirkpatrick** [MultiMC/Launcher#4682]
### Operating System

Linux

### Description of bug

Every time I try to add my Microsoft account I get the following error:
"Failed to get Minecraft access token: Error transferring https://api.minecraftservices.com/launcher/login - server replied: Forbidden"
![Screenshot from 2022-05-15 15-07-59](https://user-images.githubusercontent.com/1149447/168490064-7475f07b-c012-4b35-83a5-39f1d58247c6.png)

The account was working fine for about 1 month or so after migration - it had completed successfully and everything worked.

### Steps to reproduce

Remove account / profile 

Restart computer

Launch MultiMC

"Profiles" > "Manage Accounts" > "Add Microsoft"

Click the link / paste the token

Get the email with the second token for the "challenge" on the XBox site

Approve MultiMC in the browser

See message in browser saying it is safe to close this window

Error message appears in MultiMC 

### Suspected cause

The token was "Expired" - indicating that I need to refresh. That is when I remove the profile and problems began.
```text
    3.143 W QCoreApplication::postEvent: Unexpected null receiver
    3.144 D Task "LaunchController(0x2c3df10)" starting for the first time
    4.860 C Task "LaunchController(0x2c3df10)" failed:  "The account has expired and needs to be logged into manually again."
```

Not sure if this MultiMC, Microsoft, or a Cloudfront problem:
```text
  2.834 D Task "LaunchController(0x19a6680)" starting for the first time
    6.071 D Selected index 0 channel id "stable"
    9.606 D Task "MSAInteractive(0x1b73930)" starting for the first time
    9.606 D AuthFlow: "Logging in with Microsoft account."
    9.606 D DeviceFlow::link
    9.606 D DeviceFlow::setLinked: false
    9.931 D DeviceFlow::onDeviceAuthReplyFinished
    9.931 D DeviceFlow::onDeviceAuthReplyFinished: Tokens returned:

    9.931 D "device_code" :  "AAQABAAEAAAD--DLA3VO7QrddgJg7WevrG9-gR6kvpCT2eX8Wdrv8DJ_bpYWuh6EAIHNIqjpJNbypupXHBHv_jiRQt_53avD8Wg07BVfCewpFEE1kP_MHfxaWurZ20sH3i5XWZqj3MuqVVD8ltxrAF6gc29co1c1rQw7_k0d7dqxBkcidyIP5D_vE70YYDlY2GMSbvbM6-GcE8ey_hPGdxHNV8A7bkeGQFZBx-aHJV8Iar7tneqdlSbZy9w6t4O8YuXCsHaQgmwSDXp-NXQ9BuzevCuIsAot4tjnr2DsWl_OaqKUoELm9_dL70Dm_YXhm0RXlB1EGQZ_ku3EgPNGiaSP9Ss6xw9y-LrCiqHasHtnzDLX0OF185WbA1PBuz86UUZRFcszC5AsNEIJbEl4PStsjrUOytu5fIAA"
    9.931 D "expires_in" :  "900"
    9.931 D "interval" :  "5"
    9.931 D "message" :  "To sign in, use a web browser to open the page https://www.microsoft.com/link and enter the code MRZJH3AM to authenticate."
    9.931 D "user_code" :  "MRZJH3AM"
    9.931 D "verification_uri" :  "https://www.microsoft.com/link"
    9.932 D DeviceFlow::onDeviceAuthReplyFinished: Device auth request response
    9.932 D DeviceFlow::startPollServer: device_ and user_code expires in 900 seconds
   15.100 D PollServer::onPollTimeout: retrying
   20.100 D PollServer::onPollTimeout: retrying
   25.102 D PollServer::onPollTimeout: retrying
   30.101 D [Analytics] "Query string = v=1&tid=UA-87731965-2&cid=9ecea48b-8c8c-4039-8f78-52d2281116d1&t=screenview&ul=en-us&vp=1205x647&sr=2048x1152&aip=1&cd=Main Window&an=MultiMC&av=0.6.15-3244&cd1=1.8.0_312&cd2=64&cd3=64&cd4=64&cd5=Linux&cd6=5.13.0-41-generic&cd7=ubuntu&cd8=20.04&cm1=512&cm2=1024&cm3=15902&qt=29643"
   30.101 D PollServer::onPollTimeout: retrying
   30.171 D [Analytics] "Message sent"
   35.102 D PollServer::onPollTimeout: retrying
   40.101 D PollServer::onPollTimeout: retrying
   45.101 D PollServer::onPollTimeout: retrying
   50.100 D PollServer::onPollTimeout: retrying
   55.102 D PollServer::onPollTimeout: retrying
   60.102 D PollServer::onPollTimeout: retrying
   60.482 D DeviceFlow::onVerificationReceived: Emitting closeBrowser()
   60.482 D DeviceFlow::onVerificationReceived: Access token returned for implicit or device flow
   60.482 D DeviceFlow::onVerificationReceived: Token expires in 3600 seconds
   60.482 D DeviceFlow::setLinked: true
   60.482 D AuthFlow: "Logging in as an Xbox user."
   60.483 D First layer of XBox auth ... commencing.
   60.724 D Parsing "UToken" :
   60.724 D "UToken" is valid.
   60.725 D AuthFlow: "Getting authorization to access Xbox services."
   60.725 D Getting authorization token for  "http://xboxlive.com"
   61.170 D Parsing "Xbox" :
   61.171 D "Xbox" is valid.
   61.171 D AuthFlow: "Getting authorization to access Mojang services."
   61.171 D Getting authorization token for  "rp://api.minecraftservices.com/"
   61.437 D Parsing "Mojang" :
   61.437 D "Mojang" is valid.
   61.437 D AuthFlow: "Accessing Mojang services."
   61.438 D Getting Minecraft access token...
   61.567 W AuthRequest::onRequestError: Error 201
   61.567 W AuthRequest::onRequestError: Error string:  "Error transferring https://api.minecraftservices.com/launcher/login - server replied: Forbidden"
   61.567 W AuthRequest::onRequestError: HTTP status 403 "Forbidden"
   61.567 D "<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<TITLE>ERROR: The request could not be satisfied</TITLE>
</HEAD><BODY>
<H1>403 ERROR</H1>
<H2>The request could not be satisfied.</H2>
<HR noshade size="1px">
Request blocked.
We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.
<BR clear="all">
If you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.
<BR clear="all">
<HR noshade size="1px">
<PRE>
Generated by cloudfront (CloudFront)
Request ID: JVz53rnXTAt3OztROIzbE99ZZlijLrp6zWt1Miejprqbz63tzQggjw==
</PRE>
<ADDRESS>
</ADDRESS>
</BODY></HTML>"
   61.568 W Reply error: 201
```

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

**Comment by phit** [comment]
what is your multimc version?

**Comment by jkirkpatrick** [comment]
> ~what is your multimc version?~ found it in the log, try switching to development version

Is there a a CI server or place to get dev builds? I have too many previous scars from Qt3/4 projects in a previous life to consider that option.

**Comment by phit** [comment]
in the settings theres an option for update channel, change that to development then search for updates

**Comment by jkirkpatrick** [comment]
Updated to development version. Same error.
```text
0.001 D <> Log initialized.
    0.001 D "MultiMC 5" , (c) 2013-2021  "MultiMC Contributors"
    0.001 D Version                    :  "0.6.16-develop-3284"
    0.001 D Git commit                 :  "f45f83173662ea8d28a6d69a5312679df76d762b"
    0.001 D Git refspec                :  "refs/heads/develop"
```
... more logs ...

```text
   5.914 D Task "MSAInteractive(0x27032b0)" starting for the first time
    5.914 D AuthFlow: "Logging in with Microsoft account."
    5.921 D DeviceFlow::link
    5.921 D DeviceFlow::setLinked: false
    6.631 D DeviceFlow::onDeviceAuthReplyFinished
    6.631 D DeviceFlow::onDeviceAuthReplyFinished: Tokens returned:

    6.631 D "device_code" :  "RAQABAAEAAAD--DLA3VO7QrddgJg7Wevr1eoUVB04E0xbEWBjkPbGQf-lx7949G4zTVDclqFrBwFPRSfT3_HlAMz7pXpN1jplI4vKVtbRjCMOmfZzMzEnKuB7aLkde84JFN2z2e-1rOV-if0yiwew-POns26YdMwFth1_jxQfx8O6QiYmdgMb0bdYLc6x3iERc9dHN5kl4zPSCzVch-gILKvGxjQocTa38C7fZr5vITPIR7FFAZ_vPReYQmizxHONMaMYO9Zvfx9E7G3JlF6CV6WSYm9Vs6WZ-s8dpOoiJuATRg8PKe-L_MlxmjfFPryliRANpLT3MwPMduYxCsSENPP_SaCPxuh79POm_EA2gFceve-vItBhwQRz9DgsdV1NAgp2FWLHlyneGIv9i2uU6UAIwh5sFBdLIAA"
    6.631 D "expires_in" :  "900"
    6.631 D "interval" :  "5"
    6.632 D "message" :  "To sign in, use a web browser to open the page https://www.microsoft.com/link and enter the code UDJQ7X4E to authenticate."
    6.632 D "user_code" :  "UDJQ7X4E"
    6.632 D "verification_uri" :  "https://www.microsoft.com/link"
    6.632 D DeviceFlow::onDeviceAuthReplyFinished: Device auth request response
    6.633 D DeviceFlow::startPollServer: device_ and user_code expires in 900 seconds
   12.123 D PollServer::onPollTimeout: retrying
   17.123 D PollServer::onPollTimeout: retrying
   22.123 D PollServer::onPollTimeout: retrying
   27.123 D PollServer::onPollTimeout: retrying
   30.123 D [Analytics] "Query string = v=1&tid=UA-87731965-2&cid=9ecea48b-8c8c-4039-8f78-52d2281116d1&t=screenview&ul=en-us&vp=1205x647&sr=2048x1152&aip=1&cd=Main Window&an=MultiMC&av=0.6.16-develop-3284&cd1=1.8.0_312&cd2=64&cd3=64&cd4=64&cd5=Linux&cd6=5.13.0-41-generic&cd7=ubuntu&cd8=20.04&cm1=512&cm2=1024&cm3=15902&qt=29656"
   30.193 D [Analytics] "Message sent"
   33.123 D PollServer::onPollTimeout: retrying
   38.123 D PollServer::onPollTimeout: retrying
   43.123 D PollServer::onPollTimeout: retrying
   43.701 D DeviceFlow::onVerificationReceived: Emitting closeBrowser()
   43.701 D DeviceFlow::onVerificationReceived: Access token returned for implicit or device flow
   43.701 D DeviceFlow::onVerificationReceived: Token expires in 3600 seconds
   43.701 D DeviceFlow::setLinked: true
   43.701 D AuthFlow: "Logging in as an Xbox user."
   43.701 D First layer of XBox auth ... commencing.
   43.979 D Parsing "UToken" :
   43.979 D "UToken" is valid.
   43.979 D AuthFlow: "Getting authorization to access Xbox services."
   43.979 D Getting authorization token for  "http://xboxlive.com"
   44.474 D Parsing "Xbox" :
   44.474 D "Xbox" is valid.
   44.475 D AuthFlow: "Getting authorization to access Mojang services."
   44.475 D Getting authorization token for  "rp://api.minecraftservices.com/"
   44.697 D Parsing "Mojang" :
   44.697 D "Mojang" is valid.
   44.697 D AuthFlow: "Accessing Mojang services."
   44.698 D Getting Minecraft access token...
   44.822 W AuthRequest::onRequestError: Error 201
   44.822 W AuthRequest::onRequestError: Error string:  "Error transferring https://api.minecraftservices.com/launcher/login - server replied: Forbidden"
   44.822 W AuthRequest::onRequestError: HTTP status 403 "Forbidden"
```

**Comment by pfeiferocks** [comment]
i have the same problem. CloudFront is also blocking me 😡 

This is because my inet provider shares the ipv4 adress with multiple persons (DS-Lite) and the endpoint doesn't have an ipv6 adress so i can connect to it with my ipv6 adress 

time to setup a vpn ...

**Comment by JavaSkan** [comment]
Same problème on Windows, but also in the original website and the standard Minecraft Launcher

https://user-images.githubusercontent.com/58995237/171177555-176afdad-9b1b-4dad-b1c5-37246237bb4d.mp4

 It's in french, but you can see clearly that I'm trying to connect with a microsoft account
In Multi MC, before removing my MS account by accident, it said that the account has expired

I don't know what to do

**Comment by JavaSkan** [comment]
> contact support, we can't help you https://help.minecraft.net/hc/fr-fr/requests/new?ticket_form_id=360001225811

That's what i did

**Comment by jkirkpatrick** [comment]
This seems to be a Microsoft / Mojang issue - or possibly Cloudflare - but it has little to do with MultiMC. It took me over a week contacting MS, Mojang, and XBox to try and resolve the issue. 

I would suggest capturing HTTP 403 (Forbidden) errors in authentication, and the "expired" message in order to present something a bit more useful to the user - pointing the finger at the real culprit instead of being left to assume it's MultiMC. 

1. "Expired" - There is an issue with your login; do NOT delete the account in MultiMC. Please try to use the Minecraft launcher to login in order to fix the issue. Contacting Mojang support (https://help.minecraft.net) may be required in order to resolve your account issue. <<raw error message here>>
2. "Forbidden" - MultiMC is not able to authenticate in order to launch Minecraft. If this is your first time logging in with MultiMC, please try to use the Minecraft launcher to login in order to fix the issue. Contacting Mojang support (https://help.minecraft.net) may be required in order to resolve your account issue.  <<raw error message here>>

**Comment by JavaSkan** [comment]
@jkirkpatrick Actually i managed to fix the problem. I just switched to another internet access point (4G on my phone), and it seems that it actually worked, idk why but now it's working like before

And yeah sorry, i thought it was a MultiMC bug, but however, we have a solution now, so try it (just change your internet access point, 4G is not necessary, i think it's an IP trafic bug thing or something else)

**Comment by 1337Mishka** [comment]
Using a VPN helped me login but I am only writing this after having tried once. The issue might come back again.

**Comment by SirPotatorito** [comment]
Any progress on the issue, I'm yet to be able to log in, and as shown above I can't even log in via the offical Minecraft website. I doubt the servers are down which is the first thing that came to my mind.

**Comment by phit** [comment]
> Any progress on the issue, I'm yet to be able to log in, and as shown above I can't even log in via the offical Minecraft website. I doubt the servers are down which is the first thing that came to my mind.

If you can't even login on the official website contact Microsoft, nothing we can help you with.

**Comment by kjshrimp227** [comment]
It was the VPN being enabled for me. Simply disabled the VPN and it worked like a charm.

**Comment by Splarkszter** [comment]
When credentials expire. If i try to play MultiMC starts to spamm the damm thing trying to log-in. And got flagged as bot, now i can't log in from multi-mc due to some dumb loop in the code spamming the log-in attempts.

**Comment by MapleTheOne** [comment]
> When credentials expire. If i try to play MultiMC starts to spamm the damm thing trying to log-in. And got flagged as bot, now i can't log in from multi-mc due to some dumb loop in the code spamming the log-in attempts.

Did the same thing to me today. I just installed Windows 10 and I was trying to confirm that everything works. However, I logged into multiMC only for it to spam the service and blacklist me. I am not sure if I am permanently incapable of accessing it or temporarily.

**Comment by Splarkszter** [comment]
> > I am not sure if I am permanently incapable of accessing it or temporarily.

It is temporal. wait like 5-15 minutes and you can log into the thing again. But that loop that flags you it's dangerous.

**Comment by NotSpooky** [comment]
> When credentials expire. If i try to play MultiMC starts to spamm the damm thing trying to log-in. And got flagged as bot, now i can't log in from multi-mc due to some dumb loop in the code spamming the log-in attempts.

The same happened to me today. This is serious enough, if someone gives me pointers I can help fix it.

**Comment by Justhelpingout420** [comment]
If anyone is trying to fix this now
I had the same thing on at launcher, multi mc, prism launcher, and the Minecraft launcher all of them did the same thing.
What I did is uninstall all of them then reinstall them. ( I don't know if its necessary but I used an app called BCuninstaller)
And this completely fixed my issue
But just make sure that you don't loose anything back up anything important like instances mods or saves/worlds or anything else important you don't want to loose.
